modules/openxr/scene/openxr_render_model_manager.cpp

```diff
@@ -0,0 +1,290 @@
+}
+
+void OpenXRRenderModelManager::_update_models() {
+	OpenXRRenderModelExtension *render_model_extension = OpenXRRenderModelExtension::get_singleton();
+	ERR_FAIL_NULL(render_model_extension);
+
+	// Make a copy of our current models.
+	HashMap<RID, Node3D *> org_render_models = render_models;
+
+	// Loop through our interaction data so we add new entries.
+	TypedArray<RID> render_model_rids = render_model_extension->render_model_get_all();
+	for (const Variant &v : render_model_rids) {
+		const RID rid = v;
+		bool filter = false;
+
+		if (tracker != 0) {
+			XrPath model_path = render_model_extension->render_model_get_top_level_path(rid);
+			if (model_path != xr_path) {
+				// ignore this.
+				filter = true;
+			}
+		}
+
+		if (!filter) {
+			if (render_models.has(rid)) {
+				org_render_models.erase(rid);
+			} else {
+				// Create our container node before adding our first render model.
+				if (container == nullptr) {
+					container = memnew(Node3D);
+					add_child(container);
+				}
+
+				OpenXRRenderModel *render_model = memnew(OpenXRRenderModel);
+				render_model->set_render_model(rid);
+				container->add_child(render_model);
+				render_models[rid] = render_model;
+
+				emit_signal(SNAME("render_model_added"), render_model);
+			}
+		}
+	}
+
+	// Remove models we no longer need.
+	for (const KeyValue<RID, Node3D *> &e : org_render_models) {
+		// We sent this just before removing.
+		emit_signal(SNAME("render_model_removed"), e.value);
+
+		if (container) {
+			container->remove_child(e.value);
+		}
+		e.value->queue_free();
+		render_models.erase(e.key);
+	}
+
+	is_dirty = false;
+}
+
+void OpenXRRenderModelManager::_on_render_model_added() {
+	if (_has_filters()) {
+		// We'll update this in internal process.
+		is_dirty = true;
+	} else {
+		// No filters? Do this right away.
+		_update_models();
+	}
+}
+
+void OpenXRRenderModelManager::_on_render_model_removed() {
+	if (_has_filters()) {
+		// We'll update this in internal process.
+		is_dirty = true;
+	} else {
+		// No filters? Do this right away.
+		_update_models();
+	}
+}
+
+void OpenXRRenderModelManager::_on_render_model_top_level_path_changed(RID p_path) {
+	if (_has_filters()) {
+		// We'll update this in internal process.
+		is_dirty = true;
+	}
+}
+
+void OpenXRRenderModelManager::_notification(int p_what) {
+	// Do not run in editor!
+	if (Engine::get_singleton()->is_editor_hint()) {
+		return;
+	}
+
+	OpenXRRenderModelExtension *render_model_extension = OpenXRRenderModelExtension::get_singleton();
+	ERR_FAIL_NULL(render_model_extension);
+	if (!render_model_extension->is_active()) {
+		return;
+	}
+
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			_update_models();
+
+			render_model_extension->connect(SNAME("render_model_added"), callable_mp(this, &OpenXRRenderModelManager::_on_render_model_added));
+			render_model_extension->connect(SNAME("render_model_removed"), callable_mp(this, &OpenXRRenderModelManager::_on_render_model_removed));
```

Comment:
This one too! `OpenXRRenderModelManager::_on_render_model_removed()` should be updated to take an argument